### PR TITLE
This builds flow for injecting permanent mocks, in some cases where it was not happening earlier

### PR DESCRIPTION
**Pull request by kartikeytewari-ul**
The build fixes issue-574.

## Approach
- The existence of both `applicationContext` and `springBeanFactory` is found.
- If `applicationContext` is null, no further action is taken.
- If `applicationContext` is not null, a reflection-based search is done to identify all beans.
- Subsequently, a search is performed through the bean hierarchy to find a match for the `sourceClassName`.
- Upon locating a matching class definition, all of its constructors are identified using reflection.
- If a `no-argument constructor` is present, an object is instantiated using it.
- In the absence of a `no-argument constructor`, the first available constructor is used, with default values.
- This instantiation process yields an object for the class in question.
- Following this, the mocking flow proceeds to inject the necessary values.

### CODE REVIEW

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java [hunk 1]
```diff
@@ -483,23 +483,104 @@ public AgentCommandResponse injectMocks(AgentCommandRequest agentCommandRequest)
...
+					continue;
+				}
             }
             List<DeclaredMock> declaredMocks = mocksBySourceClass.get(sourceClassName);
 
             Map<String, List<DeclaredMock>> mocksByField = declaredMocks.stream()
                     .collect(Collectors.groupingBy(DeclaredMock::getFieldName));
 
-            Class<? extends Object> classObject = sourceClassInstance.getClass();
+			if (classObject == null) {
```
Comment (kartikeytewari-ul) [inline src/main/java/io/unlogged/AgentCommandExecutorImpl.java:576]: @artpar If both the strategies give a classObject then the first strategy will take precedence. Is that the right priority order, or should the priority be give to the value that comes from logger?

Comment (kartikeytewari-ul) [inline src/main/java/io/unlogged/AgentCommandExecutorImpl.java:576]: In line-485 if `sourceClassInstance` is found then code block from line-488 will not be called and `classObject` will be defined in line-577. So old strategy will take precedence. If in line-485 `sourceClassInstance` is not found then the code block from line-488 will run and `classObject` will be defined by new strategy. So, if both strategies work then values will be set from old strategy. Is this the right priority order?

Comment (artpar) [reply src/main/java/io/unlogged/AgentCommandExecutorImpl.java:576]: For now it's fine until we come across any issue

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java [hunk 2]
```diff
@@ -483,23 +483,104 @@ public AgentCommandResponse injectMocks(AgentCommandRequest agentCommandRequest)
...
             Object sourceClassInstance = logger.getObjectByClassName(sourceClassName);
+			
+			Class<? extends Object> classObject = null;
             if (sourceClassInstance == null) {
-                // no instance found for this class
-                // nothing to inject mocks to
-                continue;
+
+				// define applicationContext and springBeanFactory
+				if (this.springTestContextManager == null) {
```
Comment (artpar) [inline src/main/java/io/unlogged/AgentCommandExecutorImpl.java:491]: the block for spring context setup should be refactored to a separate method.

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java [hunk 3]
```diff
@@ -483,23 +483,104 @@ public AgentCommandResponse injectMocks(AgentCommandRequest agentCommandRequest)
...
+
+						Method getBeanMethod = applicationContextClass.getMethod("getBean", String.class);
+						Object bean = getBeanMethod.invoke(applicationContext, beanName);
+						applicationBean.add(bean);
+					}
+
+					// inject mocks
+					// TODO: improve search performance here
+					Class<? extends Object> classDefination = null;
+					for (Object bean: applicationBean) {
```
Comment (artpar) [inline src/main/java/io/unlogged/AgentCommandExecutorImpl.java:525]: we have the applicationBean at this point. why now lookup its class and then lookup its constructor and create the bean (instance) again ? so the object now which is created in the flow that follows its not the instance which spring is going to use to serve requests. which means no mocks.

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java (hunk 1 above)
Comment (kartikeytewari-ul) [reply src/main/java/io/unlogged/AgentCommandExecutorImpl.java:576]: okay the order looks fine then.

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java (hunk 3 above)
Comment (kartikeytewari-ul) [reply src/main/java/io/unlogged/AgentCommandExecutorImpl.java:525]: ok, got it. I have changed the flow to use the object that spring is using. I still need to do the search to match the correct bean, but have now removed the logic where constructors were getting invoked.

File: src/main/java/io/unlogged/AgentCommandExecutorImpl.java (hunk 2 above)
Comment (kartikeytewari-ul) [reply src/main/java/io/unlogged/AgentCommandExecutorImpl.java]: Done, refactored the method.

**Pull request merged**